### ISSUE TITLE
[Bug]: Loss computation with auto_mix_prec = True 

### ISSUE THREAD

**Issue by JuanPZuluaga** [speechbrain/speechbrain#1974]
### Describe the bug

Hello, 

I'm training some models with auto_mix_prec = True. But at some point, my model start getting nans. I tried to debug and found out that some tensors after forward pass end up with NaNs. I was checking online and I [found this](https://pytorch.org/docs/stable/amp.html#:~:text=Autocasting,-class%20torch.&text=Instances%20of%20autocast%20serve%20as,improve%20performance%20while%20maintaining%20accuracy)

So my question is:

Why in: https://github.com/speechbrain/speechbrain/blob/891318f5950c337bb951912bf64bd5973af7c908/speechbrain/core.py#L938

the loss computation is done outside of `with torch.autocast()` context manager? Following the PyTorch link above, it says that loss and forward pass should be inside autocasting.

thanks!

### Expected behaviour

The context manager should be able to handle the NaNs.

**Comment by AsuMagic** [comment]
Possibly related? https://github.com/speechbrain/speechbrain/issues/1803

**Comment by JuanPZuluaga** [comment]
Hi @Adel-Moumen  and @AsuMagic 

Thank you for quick reply. I'm running an encoder (conformer)-decoder transformer model, similar to the one for Librispeech. However, I'm running it for speech translation. 

Could it be related to the Conformer architecture? (I don't see this 'underflow' error in the vanilla transformer)

**Comment by Adel-Moumen** [comment]
Hmmm I don't think so. It might come from your data instead. Could you please check if you don't have outliers data (e.g., empty files?) ?

**Comment by JuanPZuluaga** [comment]
Actually, when I was debugging, I saw that in some cases is Tensors with NaNs produced by the compute_forward. I imagine they can be masked before being sent to loss computation. 

However, why this only happens after some epochs? In that, wouldn't it should happen in the first epoch?

**Comment by Adel-Moumen** [comment]
Hmmmmm when you are setting `auto_mix_prec = False` what happens ? Is it working?

**Comment by JuanPZuluaga** [comment]
Hello! I solved this issue, and it was not related to the model. 

the problem goes to the way how SpeechBrain handle the `--optimizer_step_limit` passed from CLI. 

I think this a bug. Let me describe the problem. 

- Imagine in your `hparams.yaml` file you have instantiated this scheduler: 

```
optimizer_step_limit: 1000

lr_scheduler: !new:speechbrain.nnet.schedulers.WarmCoolDecayLRSchedule
   lr: !ref <lr_adam>
   warmup: !ref <n_warmup_steps>
   cooldown: !ref <cooldown>
   total_steps: !ref <scheduler_step_limit>
   decay_every: !ref <scheduler_step_limit>*0.1

```

but then, when you call the training script you call something like this:


` python train.py hparams.yaml --optimizer_step_limit 2000` What would happen is that the initial object `lr_scheduler` would use `!ref <scheduler_step_limit> = 1000` instead of the new value passed in the CLI (ie 2000 steps). Then, during training when reaching epoch `1001`, the *LR would go below 0*, ie negative. That was my issue. The LR was going negative. 

I know this happens in when the arg is obtained from CLI but not updated everywhere: 

```
2023-05-16 03:14:36,718 - speechbrain.core - INFO - Info: optimizer_step_limit arg overridden by command line input to: 2000
```

Pablo

**Comment by Adel-Moumen** [comment]
Hey @pplantinga @Gastron, what do you think about this please?

**Comment by Gastron** [comment]
So to me it seems that this is a problem with a specific YAML file, which references the value of `optimizer_step_limit`. If optimizer_step_limit is passed on the command line, it will not be treated as a YAML override, but as one of the `run_opts`.

However, it is easy to run into this issue, because the run_opts can also be specified in YAML. Their precedence is: defaults < YAML < commandline. We cannot simply add the run_opts to overrides, as they most run_opts will not be specified in YAML, and would instead raise an error (override for a non-existing key).

Perhaps we could allow the load_hyperpyyaml to take an additional run_opts argument, which would then be applied like overrides, but would not raise an error for non-existing keys.

**Comment by pplantinga** [comment]
> Perhaps we could allow the load_hyperpyyaml to take an additional run_opts argument, which would then be applied like overrides, but would not raise an error for non-existing keys.

While I don't like the idea of adding a speechbrain-specific argument to hyperpyyaml, it could actually be nice (and sufficiently generic) to convert the current `overrides_must_match` boolean argument into a new argument, a second overrides dict that doesn't necessarily need to match. A little less cleanly, but more backwards-compatible would be to just add the non-matching overrides argument without removing `overrides_must_match`. I suppose we could add a deprecation notice on the `overrides_must_match` and remove it in the far future.

I'm not sure how hard this would be to add, as the current setup assumes all overrides will be treated in the same way. I'm also not sure who all would be able to tackle it right now.

**Comment by Gastron** [comment]
Ah yeah good point about the separation of SpeechBrain and HyperPyYAML.

Another obstacle in changing this is that it would require recipes to call load_hyperpyyaml differently, which would potentially mean changes in every recipe. Maybe it doesn't make sense to change anything.

Perhaps the best fix for this is being careful with referencing the value of optimizer_step_limit in HyperPyYAML, preferring the pattern shown by @JuanPZuluaga  above with a separate `scheduler_step_limit` argument.

**Comment by Adel-Moumen** [comment]
Thanks guys! So I guess that this issue should not be alive anymore? If you feel like that we should left it open then feel free to reopen it :-)
